api/service/projects_service.go

```diff
@@ -112,72 +118,91 @@ func (service *projectsService) save(project *models.Project) (*models.Project,
-	roleName := fmt.Sprintf("%s-%s", project.Name, "readers")
-	role, err := service.authEnforcer.UpsertRole(roleName, project.Readers)
-	if err != nil {
-		return "", err
+func (service *projectsService) updateAuthorizationPolicy(ctx context.Context, project *models.Project) error {
+	updateRequest := enforcer.NewAuthorizationUpdateRequest()
+	for _, role := range rolesWithReadOnlyAccess(project) {
+		updateRequest.UpdateRolePermissions(role, readPermissions(project))
+	}
+	if project.Administrators != nil {
+		updateRequest.UpdateRoleMembers(projectAdminRole(project), project.Administrators)
+	} else {
+		updateRequest.UpdateRoleMembers(projectAdminRole(project), []string{})
 	}
-	return role.ID, nil
+
+	for _, role := range rolesWithAdminAccess(project) {
+		updateRequest.UpdateRolePermissions(role, adminPermissions(project))
+	}
+	if project.Readers != nil {
+		updateRequest.UpdateRoleMembers(projectReaderRole(project), project.Readers)
+	} else {
+		updateRequest.UpdateRoleMembers(projectReaderRole(project), []string{})
+	}
+
+	return service.authEnforcer.UpdateAuthorization(ctx, updateRequest)
 }
 
-func (service *projectsService) upsertAdministratorsRole(project *models.Project) (string, error) {
-	roleName := fmt.Sprintf("%s-%s", project.Name, "administrators")
-	policy, err := service.authEnforcer.UpsertRole(roleName, project.Administrators)
+func (service *projectsService) filterAuthorizedProjects(ctx context.Context, projects []*models.Project,
+	user string) ([]*models.Project, error) {
+	if user == "" {
+		return nil, fmt.Errorf("authorization is enabled but user is not provided")
+	}
+
+	roles, err := service.authEnforcer.GetUserRoles(ctx, user)
 	if err != nil {
-		return "", err
-	}
-	return policy.ID, nil
-}
-
-func (service *projectsService) upsertAdministratorsPolicy(role string, project *models.Project) error {
-	subResources := fmt.Sprintf(ProjectSubResources, project.ID)
-	resource := fmt.Sprintf(ProjectResources, project.ID)
-	nameResource := fmt.Sprintf(ProjectResources, project.Name)
-	policyName := fmt.Sprintf("%s-administrators-policy", project.Name)
-	_, err := service.authEnforcer.UpsertPolicy(
-		policyName,
-		[]string{role},
-		[]string{},
-		[]string{resource, subResources, nameResource},
-		[]string{enforcer.ActionAll})
-	return err
-}
-
-func (service *projectsService) upsertReadersPolicy(role string, project *models.Project) error {
-	subResources := fmt.Sprintf(ProjectSubResources, project.ID)
-	resource := fmt.Sprintf(ProjectResources, project.ID)
-	nameResource := fmt.Sprintf(ProjectResources, project.Name)
-	policyName := fmt.Sprintf("%s-readers-policy", project.Name)
-	_, err := service.authEnforcer.UpsertPolicy(
-		policyName,
-		[]string{role},
-		[]string{},
-		[]string{resource, subResources, nameResource},
-		[]string{enforcer.ActionRead})
-	return err
+		return nil, err
+	}
+	for _, role := range roles {
+		if slices.Contains([]string{"mlp.administrator", "mlp.projects.reader"}, role) {
```

Comment:
While this is acceptable, I have the same concern as this comment: https://github.com/caraml-dev/mlp/pull/89/files#r1272876627

Can we not retrieve this from the configs already stored in Keto?

Similarly, to check membership to a project, can we retrieve the permissions from Keto instead and attempt to filter the list? It should only be 1 or 2 additional calls to Keto (as opposed to the previous way where we were making 1 call per project), right?

Reply:
Yes, we can retrieve this from Keto directly, by retrieving all the permissions. We will need to retrieve all the roles of a user, then list the permissions of each role.

The problem is that, for an mlp admin, who have all possible permissions, that means a large number of rows will be retrieved from postgres (it's roughly equals to the number of projects * the number of methods). Hence, we will run into performance issue eventually.

Reply:
That being say, there are ways to optimize the permission model such that there are less permissions that we need to deal with. So we *could* implement the above, but it does take time to design a more optimal permission model.

Reply:
> The problem is that, for an mlp admin, who have all possible permissions, that means a large number of rows will be retrieved from postgres

Can this be solved by implementing the role inheritance described [here](https://github.com/caraml-dev/mlp/pull/89#discussion_r1272913205)?

Reply:
No, it can't, because ultimately, the number of all possible permissions is the number of projects multiplied by the http methods, regardless of the number of roles.

Reply:
So the only real optimization we can do is to reduce the number of permissions itself, by perhaps mapping different http methods to the same permission eg. mlp.projects.1.write = (mlp.projects.1.post, mlp.projects.1.patch, mlp.projects.put). But even then, we will still have at the very least, two times the number of projects.

The other approach is by simply emulating glob matching, by having a permission called mlp.projects.*.write. But that has other drawbacks as described in my other comment.

Reply:
Caching role to permission matching is also a possibility - but the cache itself will be outdated whenever a new project is created, so we can't store it indefinitely.

Reply:
Hence, consider there's no real benefit of retrieving all permissions at the moment, i chose not to optimize this for the time being. But i do agree, that retrieving all permissions for a user is the more future proof solution, albeit with performance penalty.

Reply:
As for caching, the rudimentary cache introduced in https://github.com/caraml-dev/mlp/pull/87 is a read-aside cache with a max TTL of 10 mins. This may suffice.

If write-through, we need to handle the possibility of multiple MLP replicas as well.